libraries/AP_Compass/AP_Compass_Calibration.cpp

```diff
@@ -0,0 +1,243 @@
+        if (healthy(i) && use_for_yaw(i)) {
+            if (!start_calibration(i,retry,autosave,delay,autoreboot)) {
+                cancel_calibration_all();
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+void
+Compass::cancel_calibration(uint8_t i)
+{
+    _calibrator[i].clear();
+    AP_Notify::events.compass_cal_canceled = 1;
+    AP_Notify::events.initiated_compass_cal = 0;
+}
+
+void
+Compass::cancel_calibration_mask(uint8_t mask)
+{
+    for(uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
+        if((1<<i) & mask) {
+            cancel_calibration(i);
+        }
+    }
+}
+
+void
+Compass::cancel_calibration_all()
+{
+    cancel_calibration_mask(0xFF);
+}
+
+bool
+Compass::accept_calibration(uint8_t i)
+{
+    CompassCalibrator& cal = _calibrator[i];
+    uint8_t cal_status = cal.get_status();
+
+    if (cal_status == COMPASS_CAL_SUCCESS) {
+        _cal_complete_requires_reboot = true;
+        Vector3f ofs, diag, offdiag;
+        cal.get_calibration(ofs, diag, offdiag);
+        cal.clear();
+
+        set_and_save_offsets(i, ofs);
+        set_and_save_diagonals(i,diag);
+        set_and_save_offdiagonals(i,offdiag);
+
+        if (!is_calibrating()) {
+            AP_Notify::events.compass_cal_saved = 1;
+        }
+        return true;
+    } else {
+        return false;
+    }
+}
+
+bool
+Compass::accept_calibration_mask(uint8_t mask)
+{
+    for(uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
+        if ((1<<i) & mask) {
+            CompassCalibrator& cal = _calibrator[i];
+            uint8_t cal_status = cal.get_status();
+            if (cal_status != COMPASS_CAL_SUCCESS && cal_status != COMPASS_CAL_NOT_STARTED) {
+                // a compass failed or is still in progress
+                return false;
+            }
+        }
+    }
+
+    bool success = true;
+    for (uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
+        if ((1<<i) & mask) {
+            if (!accept_calibration(i)) {
+                success = false;
+            }
+        }
+    }
+
+    return success;
+}
+
+bool
+Compass::accept_calibration_all()
+{
+    return accept_calibration_mask(0xFF);
+}
+
+void
+Compass::send_mag_cal_progress(mavlink_channel_t chan)
+{
+    uint8_t cal_mask = get_cal_mask();
+
+    for (uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
+        CompassCalibrator& cal = _calibrator[i];
+
+        uint8_t& compass_id = i;
+        uint8_t cal_status = cal.get_status();
+
+        if (cal_status == COMPASS_CAL_WAITING_TO_START  ||
+            cal_status == COMPASS_CAL_RUNNING_STEP_ONE ||
+            cal_status == COMPASS_CAL_RUNNING_STEP_TWO) {
+            uint8_t completion_pct = cal.get_completion_percent();
+            uint8_t completion_mask[10];
+            Vector3f direction(0.0f,0.0f,0.0f);
+            uint8_t attempt = cal.get_attempt();
+
+            memset(completion_mask, 0, sizeof(completion_mask));
+
+            mavlink_msg_mag_cal_progress_send(
```

Comment:
why do we have the completion_mask and direction when they are always zero in this code?